### PR TITLE
fix: order by column of join table with skip and take

### PR DESCRIPTION
**Pull request by bluesh55**
related issue #3967 

Currently, the typeorm has a bug that causes mis-pagination when using join, skip, take, and orderBy in QueryBuilder at the same time. For example, suppose your data are:

**Users**

| ID | Name  |
|----|-------|
| 1  | User1 |
| 2  | User2 |
| 3  | User3 |
| 4  | User4 |

**Posts**

| ID | userId | content | order |
|----|--------|---------|-------|
| 1  | 1      | Post1   | 1     |
| 2  | 1      | Post2   | 2     |
| 3  | 2      | Post3   | 1     |
| 4  | 2      | Post4   | 2     |
| 5  | 3      | Post5   | 1     |
| 6  | 3      | Post6   | 2     |
| 7  | 4      | Post7   | 1     |
| 8  | 4      | Post8   | 2     |

If you want to join User and Post, sort by "order" on Post, and paginate with User, you can create a query builder as follows:

```typescript
const users = await connection.createQueryBuilder(User, "u")
          .innerJoinAndSelect("u.posts", "p")
          .orderBy({
            "u.id": "ASC",
            "p.order": "ASC"
          })
          .skip(0)
          .take(2)
          .getMany();
```

Is that right? And we expect this to happen.

```
[
  {
    id: 1,
    name: "User1",
    posts: [
      {
        id: 1,
        userId: 1,
        content: "Post1",
        order: 1
      },
      {
        id: 2,
        userId: 1,
        content: "Post2",
        order: 2
      }
    ]
  },
  {
    id: 2,
    name: "User2",
    posts: [
      {
        id: 3,
        userId: 2,
        content: "Post3",
        order: 1
      },
      {
        id: 4,
        userId: 2,
        content: "Post4",
        order: 2
      }
    ]
  }
]
```

But here’s the result.
```
[
  {
    id: 1,
    name: "User1",
    posts: [
      {
        id: 1,
        userId: 1,
        content: "Post1",
        order: 1
      },
      {
        id: 2,
        userId: 1,
        content: "Post2",
        order: 2
      }
    ]
  }
]
```

This bug occurs from logic that handles the simultaneous use of skip, take, join and orderBy in SelectQueryBuilder.

In this case, the main table must first be looked up to perform a pagination. Next, you must perform a second query that actually joins using the data from the main table where the pagination was performed.

[SelectQueryBuilder](https://github.com/typeorm/typeorm/blob/master/src/query-builder/SelectQueryBuilder.ts#L1931) also performs the first query for the pagination. However, if the column of the join table is given to the orderBy function, it will affect the first query. And this causes the pagination bug.

So I added a guard so that only the columns of the main table can be used for the first query. It works well and passes all the tests. But I don’t know exactly what other side effects there are.

### CODE REVIEW

**Comment by bluesh55** [comment]
Is the review in progress? 👀

**Comment by imnotjames** [comment]
What are the queries generated before and after?

File: src/query-builder/SelectQueryBuilder.ts [hunk 1]
```diff
@@ -2010,14 +2010,20 @@ export class SelectQueryBuilder<Entity> extends QueryBuilder<Entity> implements
...
+
+                    return "";
                 } else {
                     if (this.expressionMap.selects.find(select => select.selection === orderCriteria || select.aliasName === orderCriteria))
                         return this.escape(parentAlias) + "." + orderCriteria;
 
                     return "";
                 }
             })
+            .filter(Boolean)
```
[the remark this answers is not in the document]
Comment (bluesh55) [reply src/query-builder/SelectQueryBuilder.ts:2069]: If the callback is not passed to `filter` function, it returns 'Uncaught TypeError: undefined is not a function' error.
I'm using nodejs v14.16.1

**Comment by bluesh55** [comment]
@imnotjames 

Before and after queries are same. There is only a difference in logic that processes data.

```sql
SELECT `u`.`id` AS `u_id`, `u`.`name` AS `u_name`, `p`.`id` AS `p_id`, `p`.`content` AS `p_content`, `p`.`order` AS `p_order`, `p`.`userId` AS `p_userId`
FROM `user` `u`
INNER JOIN `post` `p` ON `p`.`userId`=`u`.`id`
ORDER BY `u`.`id` DESC, `p`.`order` DESC
```

**Comment by pleerock** [comment]
There is a difference in generated `SQL` queries.

before:

```sql
SELECT DISTINCT "distinctAlias"."u_id" as "ids_u_id", "distinctAlias"."u_id", "distinctAlias"."p_order" FROM (SELECT "u"."id" AS "u_id", "u"."name" AS "u_name", "p"."id" AS "p_id", "p"."content" AS "p_content", "p"."order" AS "p_order", "p"."userId" AS "p_userId" FROM "user" "u" INNER JOIN "post" "p" ON "p"."userId"="u"."id") "distinctAlias" ORDER BY "distinctAlias"."u_id" DESC, "distinctAlias"."p_order" DESC, "u_id" ASC LIMIT 5
SELECT "u"."id" AS "u_id", "u"."name" AS "u_name", "p"."id" AS "p_id", "p"."content" AS "p_content", "p"."order" AS "p_order", "p"."userId" AS "p_userId" FROM "user" "u" INNER JOIN "post" "p" ON "p"."userId"="u"."id" WHERE "u"."id" IN (10, 10, 9, 9, 8) ORDER BY "u"."id" DESC, "p"."order" DESC
```

after:

```sql
SELECT DISTINCT "distinctAlias"."u_id" as "ids_u_id", "distinctAlias"."u_id" FROM (SELECT "u"."id" AS "u_id", "u"."name" AS "u_name", "p"."id" AS "p_id", "p"."content" AS "p_content", "p"."order" AS "p_order", "p"."userId" AS "p_userId" FROM "user" "u" INNER JOIN "post" "p" ON "p"."userId"="u"."id") "distinctAlias" ORDER BY "distinctAlias"."u_id" DESC, "u_id" ASC LIMIT 5
SELECT "u"."id" AS "u_id", "u"."name" AS "u_name", "p"."id" AS "p_id", "p"."content" AS "p_content", "p"."order" AS "p_order", "p"."userId" AS "p_userId" FROM "user" "u" INNER JOIN "post" "p" ON "p"."userId"="u"."id" WHERE "u"."id" IN (10, 9, 8, 7, 6) ORDER BY "u"."id" DESC, "p"."order" DESC
```

as you can see `SELECT DISTINCT ... "distinctAlias"."p_order"` and `ORDER BY ... "distinctAlias"."p_order" DESC` were removed.

Without "p_order" being using in `ORDER BY` of `DISTINCT` query, `ids` returned by `DISTINCT` ids won't be correct - order isn't considered at all. And `ORDER BY` in the second query won't affect anything except order in a given `ids`. But the problem that returned `ids` already are invalid, because order was't considered and their number were limited by `LIMIT` criteria. 

Your test is too trivial and doesn't check order correctness.

I have to close this PR because fix is invalid.

**Pull request closed without merging**